csesoc/Circles
# feature(global): auth?!?! USER AUTHENTICATION?!?!?!?!

todo after merge (martin didnt want me to create more diff):
- **add a logout button on degree-wizard**
- react query key refactoring
- docstrings (will do this after approvals as itll cause annoying changes)
- tests? tests? tests? lol no later
- couple our session storage more closely with the csesoc auth
- remove the last few user axios requests from certain fronteng pages like SettingsMenu
- do a lil user route refactoring so they actually all directly use the database helpers i have created
- other stuff i have noted in minutes
- anything that is `TODO-OLLI(pm)` (pm stands for post auth merge)
- setup docker containers for production
  - this involves also dropping the redis stack to not include gui
- confluence docs baby
- figure out token batch deleting
- can probably do some preliminary error checking on the success login page for expected query params

# Changes (will format nicer in abit)

## backend changes
- removed nodemon from backend docker, in favour of uvicorn watch
- removed init-mongo from the main stack, in favour of just doing this step before the uvicorn start
- added new flags for runserver.py
- refactored the database code to all be in one spot
  - see backend/server/db
  - moved connections into here as well
  - created helper functions ontop the user related queries
  - dont clear users anymore on startup
  - cleaned up mongo setup and handling in general
- refactored the user database
  - now you have invariants
  - also has full mongo validation for all three user related collections
  - users now are either setup or not, no default users anymore
- added a bunch of request and validation helpers for the oidc auth code flow
- added redis and built out a session storage layer ontop of oidc
  - look in auth_utility/sessions
  - includes a set of interface functions that abstract the db helpers
  - supports guest sessions and fedauth (oidc) sessions
  - redis maps session token -> sid & uid
    - and is not persisted
  - mongo collections map;
    - refresh token -> sid
    - sid -> session info
    - uid -> user info
- created some middlewares for getting uid into routes
- converted all user routes to use the uid middleware
- underwrote the user routes to use adapters for now
  - for converting the db models into the TypedDicts
  - will do further conversion and remove the TypedDicts after merge
- summer term is false by default again
- added lifespan manager for future use (could probs remove)
- now requires a sessionsdb env file too
- changed workspace vscode settings so not auto run pytests
- changed run_app.py to use new docker containers
  - also made its environment more explicit. Before `load_dotenv` was just modifying the global environment, meaning the `get_x_env()` calls were a bit redundant
- broke tests (half way through fixing)
  - they atleast run now locally

## frontend changes
- now have token boundaries which handle redirecting based on user state
- added a identity provider which handles refreshing of tokens on page load
  - also auto refreshes just based on interval, other solutions have been explored and will continue to be
- added logout page
- reworked login page
- moved identity and token into its own redux slice
  - SHOULD NOT BE REDUX PERSISTED
- created useToken hook ontop of useSelector to ensure a defined token
  - also has a mode that allows undefined, still better than selector
- converted all request helpers to take in token as param
- removed any google auth related stuff
- converted all queries to use token
- changed where some links go (course-selector is now the first page for setup users)
- added degree wizard as a topbar page since it should be real page now
  - hamburger menu coming soon
- changed up degree wizard to work with new setup states
  - thus removed isComplete field from users
- landing page also now knows where to redirect you too based on isSetup
- error404 page dont show header anymore (since the cart required auth)
- created the required request helpers for auth
- created a withAuthorization request helper for formatting token header
- deleted some dead request helpers
- broke tests



roast me, burn me at the stake

(dm me if u need env files (you will need new env files), ill distribute soon)
